src/alto-ui/Form/TagInput/TagInput.js

```diff
@@ -0,0 +1,209 @@
+      const val = value;
+      if (val) {
+        if (e.key === 'Enter' && match()) {
+          handleNewTag(val);
+        }
+        return undefined;
+      }
+      if (e.key === 'ArrowLeft') return movePositionLeft();
+      if (e.key === 'ArrowRight') return movePositionRight();
+      if (e.key === 'Backspace' && tags.length && position !== 0) {
+        const indexToRemove = position === null ? tags.length - 1 : position - 1;
+        const newTags = tags.filter((_, index) => index !== indexToRemove);
+        onChange(newTags);
+        if (position !== null) setPosition(indexToRemove);
+      }
+      return undefined;
+    }
+
+    function handleClickTag(e, tag, index) {
+      if (isTagActive(tag)) {
+        removeFromSelection(tag);
+      } else {
+        const tagsToAdd =
+          e.shiftKey && selection.length
+            ? sliceArr(tags, tags.findIndex(isTagEqual(selection.slice(-1)[0])), index)
+            : [tag];
+        addToSelection(...tagsToAdd);
+      }
+    }
+
+    const input = (
+      <input
+        ref={refs.input}
+        type="text"
+        className={bemClass('TagInput__input', {
+          empty: !value,
+          valid: match(),
+        })}
+        value={value}
+        onChange={e => setValue(e.target.value)}
+        size={value ? value.length : 0}
+        onFocus={resetSelection}
+        onBlur={() => {
+          resetValue();
+          resetPosition();
+        }}
+        onKeyDown={handleKeyDownInput}
+        disabled={disabled}
+      />
+    );
+
+    return (
+      <div ref={refs.main} className={bemClass('TagInput', {}, className)}>
+        <div className="TagInput__tags-list TagInput__tags-list--value" ref={refs.value}>
+          <button
+            disabled={disabled}
+            className="TagInput__button"
+            onClick={() => setPosition(null)}
+          />
+          {tags.map((tag, index) => (
+            <Fragment key={getTagValue(tag)}>
+              {position === index ? input : <div style={{ width: 1 }} />}
+              <div className="TagInput__tag TagInput__tag--value">
+                {renderTag(tag, {
+                  rounded,
+                  active: isTagActive(tag),
+                  disabled,
+                  onClick: e => handleClickTag(e, tag, index),
+                })}
+              </div>
+            </Fragment>
+          ))}
+          {position === null || position === undefined ? input : <div style={{ width: 1 }} />}
+        </div>
+        {!disabled && state.isActive && !!remainingTags.length && (
+          <div className="TagInput__tags-list">
+            {remainingTags.map(tag => (
+              <div key={getTagValue(tag)} className="TagInput__tag">
+                {renderTag(tag, {
+                  disabled: !getTagValue(tag).includes(value),
+                  onClick: () => {
+                    handleNewTag(tag);
+                    setTimeout(() => setPosition(null), 0);
```

Comment:
`setTimeout(() => setPosition(null), 0);` neccessary ?

Reply:
I think the idea was to pospone the setState after the render (at the end of the stack). I will check if it is mandatory.